### Import PDF Document 

We're going to pretend we're nutrition students at the University of Hawai'i, reading through the open-source PDF textbook [*Human Nutrition: 2020 Edition*](https://pressbooks.oer.hawaii.edu/humannutrition2/).

There are several libraries to open PDFs with Python but I found that [PyMuPDF](https://github.com/pymupdf/pymupdf) works quite well in many cases.

First we'll download the PDF if it doesn't exist.

In [2]:
# Download PDF file
import os
import requests

# Get PDF document
pdf_path = "human-nutrition-text.pdf"

# Download PDF if it doesn't already exist
if not os.path.exists(pdf_path):
  print("File doesn't exist, downloading...")
  # The URL of the PDF you want to download
  url = "https://pressbooks.oer.hawaii.edu/humannutrition2/open/download?type=pdf"
  # The local filename to save the downloaded file
  filename = pdf_path
  # Send a GET request to the URL
  response = requests.get(url)
  # Check if the request was successful
  if response.status_code == 200:
      # Open a file in binary write mode and save the content to it
      with open(filename, "wb") as file:
          file.write(response.content)
      print(f"The file has been downloaded and saved as {filename}")
  else:
      print(f"Failed to download the file. Status code: {response.status_code}")
else:
  print(f"File {pdf_path} exists.")

File doesn't exist, downloading...
The file has been downloaded and saved as human-nutrition-text.pdf


PDF acquired!

We can import the pages of our PDF to text by first defining the PDF path and then opening and reading it with PyMuPDF (`import fitz`).

In [7]:
# Requires !pip install PyMuPDF, see: https://github.com/pymupdf/pymupdf
import fitz
from tqdm.auto import tqdm # for progress bars, requires !pip install tqdm 
import random

def text_formatter(text: str) -> str:
    """Performs minor formatting on text."""
    cleaned_text = text.replace("\n", " ").strip() # note: this might be different for each doc (best to experiment)
    # Other potential text formatting functions can go here
    return cleaned_text

# Open PDF and get lines/pages
# Note: this only focuses on text, rather than images/figures etc
def open_and_read_pdf(pdf_path: str) -> list[dict]:
    """
    Opens a PDF file, reads its text content page by page, and collects statistics.

    Parameters:
        pdf_path (str): The file path to the PDF document to be opened and read.

    Returns:
        list[dict]: A list of dictionaries, each containing the page number
        (adjusted), character count, word count, sentence count, token count, and the extracted text
        for each page.
    """
    doc = fitz.open(pdf_path)  # open a document
    pages_and_texts = []
    for page_number, page in tqdm(enumerate(doc)):  # iterate the document pages
        text = page.get_text()  # get plain text encoded as UTF-8
        text = text_formatter(text)
        pages_and_texts.append({"page_number": page_number - 41,  # adjust page numbers since our PDF starts on page 42
                                "page_char_count": len(text),
                                "page_word_count": len(text.split(" ")),
                                "page_sentence_count_raw": len(text.split(". ")),
                                "page_token_count": len(text) / 4,  # 1 token = ~4 chars, see: https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them
                                "text": text})
    return pages_and_texts

pages_and_texts = open_and_read_pdf(pdf_path=pdf_path)
pages_and_texts[:2]

0it [00:00, ?it/s]

[{'page_number': -41,
  'page_char_count': 29,
  'page_word_count': 4,
  'page_sentence_count_raw': 1,
  'page_token_count': 7.25,
  'text': 'Human Nutrition: 2020 Edition'},
 {'page_number': -40,
  'page_char_count': 0,
  'page_word_count': 1,
  'page_sentence_count_raw': 1,
  'page_token_count': 0.0,
  'text': ''}]

In [8]:
import random
random.sample(pages_and_texts, k=2)

[{'page_number': 882,
  'page_char_count': 1328,
  'page_word_count': 212,
  'page_sentence_count_raw': 13,
  'page_token_count': 332.0,
  'text': 'Treatment for lead poisoning includes removing the child from  the source of contamination and extracting lead from the body.  Extraction may involve chelation therapy, which binds with lead  so it can be excreted in urine. Another treatment protocol, EDTA  therapy,  involves  administering  a  drug  called  ethylenediaminetetraacetic  acid  to  remove  lead  from  the  bloodstream of patients with levels greater than 45 mcg/dL.9  Fortunately, lead toxicity is highly preventable. It involves  identifying potential hazards, such as lead paint and pipes, and  removing them before children are exposed to them.  Learning Activities  Technology Note: The second edition of the Human  Nutrition Open Educational Resource (OER) textbook  features interactive learning activities.  These activities are  available in the web-based textbook and not avai

### Get some stats on the text

This means that the model has been trained in ingest and turn into embeddings texts with 384 tokens (1 token ~= 4 characters ~= 0.75 words).

In [9]:
import pandas as pd

df = pd.DataFrame(pages_and_texts)
df.describe().round(2)

page_number  page_char_count  page_word_count  page_sentence_count_raw  \
count      1208.00          1208.00          1208.00                  1208.00   
mean        562.50          1148.00           199.50                    10.52   
std         348.86           560.38            95.83                     6.55   
min         -41.00             0.00             1.00                     1.00   
25%         260.75           762.00           134.00                     5.00   
50%         562.50          1231.50           216.00                    10.00   
75%         864.25          1603.50           272.00                    15.00   
max        1166.00          2308.00           430.00                    39.00   

       page_token_count  
count           1208.00  
mean             287.00  
std              140.10  
min                0.00  
25%              190.50  
50%              307.88  
75%              400.88  
max              577.00

### Further text processing (splitting pages into sentences)

The ideal way of processing text before embedding it is still an active area of research.

A simple method I've found helpful is to break the text into chunks of sentences.

As in, chunk a page of text into groups of 5, 7, 10 or more sentences (these values are not set in stone and can be explored).

But we want to follow the workflow of:

`Ingest text -> split it into groups/chunks -> embed the groups/chunks -> use the embeddings`

Some options for splitting text into sentences:

1. Split into sentences with simple rules (e.g. split on ". " with `text = text.split(". ")`, like we did above).
2. Split into sentences with a natural language processing (NLP) library such as [spaCy](https://spacy.io/) or [nltk](https://www.nltk.org/).

Why split into sentences?

* Easier to handle than larger pages of text (especially if pages are densely filled with text).
* Can get specific and find out which group of sentences were used to help within a RAG pipeline.

> **Resource:** See [spaCy install instructions](https://spacy.io/usage). 

Let's use spaCy to break our text into sentences since it's likely a bit more robust than just using `text.split(". ")`. 

In [11]:
from spacy.lang.en import English # see https://spacy.io/usage for install instructions

nlp = English()

# Add a sentencizer pipeline, see https://spacy.io/api/sentencizer/ 
nlp.add_pipe("sentencizer")

# Create a document instance as an example
doc = nlp("This is a sentence. This another sentence.")
assert len(list(doc.sents)) == 2

# Access the sentences of the document
list(doc.sents)

[This is a sentence., This another sentence.]

let's run our small sentencizing pipeline on our pages of text.

In [12]:
for item in tqdm(pages_and_texts):
    item["sentences"] = list(nlp(item["text"]).sents)
    
    # Make sure all sentences are strings
    item["sentences"] = [str(sentence) for sentence in item["sentences"]]
    
    # Count the sentences 
    item["page_sentence_count_spacy"] = len(item["sentences"])

  0%|          | 0/1208 [00:00<?, ?it/s]

In [13]:
# Inspect an example
random.sample(pages_and_texts, k=1)

[{'page_number': 246,
  'page_char_count': 1159,
  'page_word_count': 215,
  'page_sentence_count_raw': 9,
  'page_token_count': 289.75,
  'text': 'Image by  Allison  Calabrese /  CC BY 4.0  starches yielding more energy than insoluble fibers. Since dietary  fiber is digested much less in the gastrointestinal tract than other  carbohydrate types (simple sugars, many starches) the rise in blood  glucose after eating them is less, and slower. These physiological  attributes of high-fiber foods (i.e. whole grains) are linked to a  decrease in weight gain and reduced risk of chronic diseases, such  as Type 2 diabetes and cardiovascular disease.  Figure 4.9 Overview of Carbohydrate Digestion  A Carbohydrate Feast  You are at a your grandma’s house for family dinner and you just  consumed kalua pig, white rice, sweet potatoes, mac salad, chicken  long rice and a hot sweet bread roll dripping with butter. Less  than an hour later you top it off with a slice of haupia pie and  then lie down on

In [14]:
df = pd.DataFrame(pages_and_texts)
df.describe().round(2)

page_number  page_char_count  page_word_count  page_sentence_count_raw  \
count      1208.00          1208.00          1208.00                  1208.00   
mean        562.50          1148.00           199.50                    10.52   
std         348.86           560.38            95.83                     6.55   
min         -41.00             0.00             1.00                     1.00   
25%         260.75           762.00           134.00                     5.00   
50%         562.50          1231.50           216.00                    10.00   
75%         864.25          1603.50           272.00                    15.00   
max        1166.00          2308.00           430.00                    39.00   

       page_token_count  page_sentence_count_spacy  
count           1208.00                    1208.00  
mean             287.00                      10.32  
std              140.10                       6.30  
min                0.00                       0.00  
25%              190.50                       5.00  
50%              307.88                      10.00  
75%              400.88                      15.00  
max              577.00                      28.00

### Chunking our sentences together

Let's take a step to break down our list of sentences/text into smaller chunks.

As you might've guessed, this process is referred to as **chunking**.

Why do we do this?

1. Easier to manage similar sized chunks of text.
2. Don't overload the embedding models capacity for tokens (e.g. if an embedding model has a capacity of 384 tokens, there could be information loss if you try to embed a sequence of 400+ tokens).
3. Our LLM context window (the amount of tokens an LLM can take in) may be limited and requires compute power so we want to make sure we're using it as well as possible.

Something to note is that there are many different ways emerging for creating chunks of information/text.

For now, we're going to keep it simple and break our pages of sentences into groups of 10 (this number is arbitrary and can be changed, I just picked it because it seemed to line up well with our embedding model capacity of 384).

On average each of our pages has 10 sentences.

And an average total of 287 tokens per page.

So our groups of 10 sentences will also be ~287 tokens long.

This gives us plenty of room for the text to embedded by our `all-mpnet-base-v2` model (it has a capacity of 384 tokens).

To split our groups of sentences into chunks of 10 or less, let's create a function which accepts a list as input and recursively breaks into down into sublists of a specified size.